### PR TITLE
Faster implementation of fANOVA

### PR DESCRIPTION
**Pull request by not522**
## Motivation
This PR refactors the implementation of `FanovaImportanceEvaluator` to improve speed.

## Description of the changes
- Adding `lru_cache` to node information getters
- Discarding NumPy's `any`
- Reducing the number of cardinality calculations

## Benchmark
I ran a benchmark using the optuna-fast-fanova's script. https://github.com/optuna/optuna-fast-fanova/blob/main/tools/benchmark.py

| n_trials | n_params | n_trees | Optuna (master) | Optuna (PR) | optuna-fast-fanova |
| -------- | -------- | ------- | --------------- | ------------------ | ------------------ |
| 100 | 2 | 64 | 1.219s | 0.844s (-30.8%) | 0.111s (-90.9%) |
| 100 | 8 | 64 | 1.611s | 1.031s (-36.0%) | 0.170s (-89.5%) |
| 100 | 32 | 64 | 1.763s | 1.245s (-29.4%) | 0.438s (-75.2%) |
| 1000 | 2 | 64 | 51.800s | 24.699s (-52.3%) | 1.359s (-97.4%) |
| 1000 | 8 | 64 | 79.404s | 34.244s (-56.9%) | 2.342s (-97.0%) |
| 1000 | 32 | 64 | 58.464s | 26.081s (-55.4%) | 3.464s (-94.1%) |

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@knshnb @Alnusjaponica Could you review this PR, please?

**Review by Alnusjaponica — APPROVED** [review]
I confirmed that this change makes fANOVA faster with provided benchmark code. LGTM.

**Comment by Alnusjaponica** [comment]
@knshnb Could you take another look?

File: optuna/importance/_fanova/_tree.py [hunk 1]
```diff
@@ -120,19 +121,24 @@ def _get_marginalized_statistics(self, feature_vector: numpy.ndarray) -> Tuple[f
                     continue
 
                 # If subtree starting from node splits on an active feature, push both child nodes.
-                if (active_features & self._subtree_active_features[node_index]).any():
+                if any(self._subtree_active_features[node_index][active_features].tolist()):
```
Comment (knshnb) [inline optuna/importance/_fanova/_tree.py:124]: We can probably add a comment on why we use `any` instead of `numpy.ndarray.any`.

**Review by knshnb — COMMENTED** [review]
I'm very sorry for the delayed response🙇  I left a small comment and let us discuss other details in person.

File: optuna/importance/_fanova/_tree.py [hunk 2]
```diff
@@ -120,19 +121,24 @@ def _get_marginalized_statistics(self, feature_vector: numpy.ndarray) -> Tuple[f
...
+            active_leaf_search_spaces.append(search_spaces)
 
         statistics = self._statistics[node_indices]
         values = statistics[:, 0]
         weights = statistics[:, 1]
+        active_features_cardinalities = numpy.asarray(active_leaf_search_spaces)
+        active_features_cardinalities = (
+            active_features_cardinalities[:, :, 1] - active_features_cardinalities[:, :, 0]
+        )
+        active_features_cardinalities = numpy.prod(active_features_cardinalities, axis=1)
```
Comment (hvy) [inline optuna/importance/_fanova/_tree.py:141]: We could introduce a `_get_cardinality_batched`, or generalize `_get_cardinality` to handle these cases. Very optional. (And thanks a lot for the PR!)

**Review by knshnb — APPROVED** [review]
I confirmed replacing `numpy.ndarray.any` to `tolist()` and `any` yields about 22% performance gain in my environment. Let me merge this PR after the comments are addressed.

**Comment by not522** [comment]
Thank you for reviews. I addressed your comments.

**Review by knshnb — APPROVED** [review]
LGTM!

**Pull request merged**